newrelic/python_agent.go

```diff
@@ -0,0 +1,71 @@
+package newrelic
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/buildpacks/libcnb"
+	"github.com/paketo-buildpacks/libpak"
+	"github.com/paketo-buildpacks/libpak/bard"
+	"github.com/paketo-buildpacks/libpak/effect"
+	"github.com/paketo-buildpacks/libpak/sherpa"
+)
+
+type PythonAgent struct {
+	BuildpackPath    string
+	ApplicationPath  string
+	Executor         effect.Executor
+	LayerContributor libpak.DependencyLayerContributor
+	Logger           bard.Logger
+}
+
+func NewPythonAgent(applicationPath string, buildpackPath string, dependency libpak.BuildpackDependency, cache libpak.DependencyCache) (PythonAgent, libcnb.BOMEntry) {
+	contributor, entry := libpak.NewDependencyLayer(dependency, cache, libcnb.LayerTypes{
+		Launch: true,
+	})
+	return PythonAgent{
+		ApplicationPath:  applicationPath,
+		BuildpackPath:    buildpackPath,
+		Executor:         effect.NewExecutor(),
+		LayerContributor: contributor,
+	}, entry
+}
+
+func (p PythonAgent) Contribute(layer libcnb.Layer) (libcnb.Layer, error) {
+	p.LayerContributor.Logger = p.Logger
+
+	p.Logger.Bodyf("Installing to %s", layer.Path)
+
+	fmt.Println("Installing New Relic Python Agent using pip3")
+	if err := p.Executor.Execute(effect.Execution{
+		Command: "pip3",
+		Args:    []string{"install", "newrelic==" + os.Getenv("NEWRELIC_PYTHON_VERSION")},
+		Dir:     p.ApplicationPath,
+		Stdout:  p.Logger.InfoWriter(),
+		Stderr:  p.Logger.InfoWriter(),
+	}); err != nil {
+		return libcnb.Layer{}, fmt.Errorf("unable to run pip install\n%w", err)
+	}
+
+	file := filepath.Join(p.BuildpackPath, "resources", "newrelic.ini")
```

Comment:
I think this might need to be user customizable. At least to set the license key?

If not, you can keep doing it this way. If it needs to be user customizable you have a couple of options.

1. You can set up a binding where the user can pass in the file. Bindings are handled like secrets, so they can contain sensitive info like a license key.
2. You could pass in just the sensitive bits like a license key through a binding (it's basically just key/value data the user can specify, a whole file or just a string, it doesn't matter). This would be a bit more work if there are a lot of values to pass through. You could then either stick with an unmodifiable default or you could allow the user to volume map a file into a specific location at which the buildpack would look and then you can copy that custom file into the app (you don't want to do this if the file has secrets though because it could result in those secrets becoming part of the image which is very bad).
3. You can also go with a default *if* the agent already has a way for users to override those values, like a standard env variable to set the license key or other functionality. Although what we would recommend is still using a binding for the secret information, since bindings are generally handled in more secure ways that env variables.

Let me know if you have more questions about this.

Reply:
By providing the newrelic.ini we ensure that app runs, if user sets the NR License key and app name as ENV's.
However, agree that best course action is to use bindings, are you able to link me to any examples?

Reply:
The AppDynamics buildpack does some stuff like this. It has a default set of configuration files that [get copied into the image at build time](https://github.com/paketo-buildpacks/appdynamics/blob/main/appd/java_agent.go#L140). There's nothing sensitive in those files though, there's a [helper](https://github.com/paketo-buildpacks/appdynamics/blob/main/helper/properties.go) like there is for this buildpack which maps bindings to env variables at run time (which is the recommended way to handle secrets, binding at runtime).

Reply:
Will now be checking if newrelic package exists somewhere in /layers/paketo-buildpacks_pip-install/packages/ and will fail if does not, if exists it will then check for the existence of newrelic.ini and copy from resources if one does not exists